lua/bufferline/state.lua

```diff
@@ -231,5 +220,34 @@ function state.restore_buffers(buffer_data)
   end
 end
 
+--- The `icons` for a particular activity.
+--- @param activity bufferline.buffer.activity.name
+--- @see bufferline.options.icons
+--- @return bufferline.options.icons.buffer
+function state.icons(bufnr, activity)
+  local activity_lower = activity:lower()
+  local icons = options.icons()
+
+  --- @type bufferline.options.icons.state
+  local activity_icons = utils.tbl_remove_key(icons, activity_lower, {})
+
+  --- @type bufferline.options.icons.buffer
+  local buffer_icons = utils.setfallbacktable(activity_icons, icons)
+
+  for option, enabled in pairs {
+    modified = buf_get_option(bufnr, 'modified'),
+    pinned = state.get_buffer_data(bufnr).pinned,
+  } do
+    if enabled then
+      buffer_icons = utils.setfallbacktable(
+        utils.tbl_remove_key(activity_icons, option, {}),
+        utils.setfallbacktable(utils.tbl_remove_key(icons, option, {}), buffer_icons)
+      )
+    end
+  end
```

Comment:
This is where we adhere the precedence I noted in the PR body. Instead of looping over tables and sending a new table of resolved values, I opted for using Lua's `__index` metamethod since it's ~20% faster than the alternative `vim.tbl_deep_extend` or `for k, v in pairs(…) do if v ~= nil then foo[k] = v` options.